Add ipynb support to R Grader (#11941)
PrairieLearn/PrairieLearn: graders/r/README.md

and more. We also include a few command-line applications such as `git` and `sqlite3`.

## File Support

This grader primarily focuses on `.R` files for both student submissions and test files. The `.R` file can be named anything, provided it uses the same name in the test files. Additionally, `.ipynb` notebooks are supported for student submissions (not test files) with the following caveats:

- the student code must be written in a notebook called `student.ipynb`

> - the answer code must be in a single cell of the notebook that starts with `#R`

Reviewer: This and a number of other places in this PR need to be updated to reflect that this is now `#grade`.

Author: Oops, missed them, fixed in 12ab7fa

and more. We also include a few command-line applications such as `git` and `sqlite3`.

## File Support

This grader primarily focuses on `.R` files for both student submissions and test files. The `.R` file can be named anything, provided it uses the same name in the test files. Additionally, `.ipynb` notebooks are supported for student submissions (test files must still be `.R` files) with the following caveats:

- the student code must be written in a notebook called `student.ipynb`
- the answer code must be in a single cell of the notebook that starts with the correct `ipynb_key` (see below)
- the test files use `/grade/student/student.R` (not `/student.ipynb`)

## ipynb_key

When grading an `ipynb` file you need to tell the autograder which cell to grade using a specific key. The default is `#grade` as in the example question. This can be configured using the environment variable in `info.json`

```json

> "externalGradingOptions": {

Author: Fixed 068fa01

    "enabled": true,
    "image": "prairielearn/grader-r",
    "timeout": 20,
    "environment": {

>       "ipynb_key": "#student_code"

Author: Updated in ea46f2e

and more. We also include a few command-line applications such as `git` and `sqlite3`.

## File Support

This grader primarily focuses on `.R` files for both student submissions and test files. The `.R` file can be named anything, provided it uses the same name in the test files. Additionally, `.ipynb` notebooks are supported for student submissions (test files must still be `.R` files) with the following caveats:

- the student code must be written in a notebook called `student.ipynb`
- the answer code must be in a single cell of the notebook that starts with the correct `IPYNB_KEY` (see below)
- the test files use `/grade/student/student.R` (not `/student.ipynb`)

## Customizing the grading key

> When grading an `ipynb` file you need to tell the autograder which cell to grade using a specific key. This key is a comment that must appear at the start of the student solution cell. All other cells will be ignored by the grader. Given it is a comment, the key must start with a hashtag (`#`). An example cell could be as follows:

Reviewer: Pedantic nit, a hashtag is the combination of `#` and the following characters.  (or If you must call it something, "hash".

```suggestion
When grading an `ipynb` file, you need to tell the autograder which cells to grade using a specific key, which is a comment starting with `#`. All other cells will be ignored by the grader. An cell using the default key of `#grade` could be as follows:
```

Author: ```suggestion
When grading an `ipynb` file, you need to tell the autograder which cells to grade using a specific key, which is a comment starting with `#`. All other cells will be ignored by the grader. A cell using the default key of `#grade` could be as follows:
```

```r
#grade (write your code in this cell and DO NOT DELETE THIS LINE)
answer <- ...
```

> In this example, `#grade` is the key. Anything after the whitespace is optional, but recommended to remind students to not modify this comment. The default key is `#grade`, but it can be configured using the `IPYNB_KEY` environment variable in `info.json`:

Reviewer: ```suggestion
Any text following the key is optional but recommended to remind students to not modify this comment. A custom key can be configured using the `IPYNB_KEY` environment variable in `info.json`:
```
